src/crawlers/basic_crawler.js

```diff
@@ -302,21 +302,23 @@ class BasicCrawler {
         await this._loadHandledRequestCount();
 
         this.isRunningPromise = this.autoscaledPool.run();
-        this.stats.startLogging();
+        await this.stats.startCapturing();
+
         try {
             await this.isRunningPromise;
         } finally {
             if (this.useSessionPool) {
                 await this.sessionPool.teardown();
             }
 
-            this.stats.stopLogging();
+            await this.stats.stopCapturing();
             const finalStats = this.stats.getCurrent();
             this.log.info('Final request statistics:', finalStats);
         }
     }
 
     async _pauseOnMigration() {
+        await this.stats.persistState();
         await this.autoscaledPool.pause(SAFE_MIGRATION_WAIT_MILLIS)
```

Comment:
The persistState should go after the `pause` resolves, because that's when all the in-progress requests have finished.

The before-pause persistence will be triggered by the `events.on` handler in Statistics.

Reply:
was about to ask this. it shouldn't be after the requestList persist though, right?

Reply:
I think you can fire them both in parallel to speed it up a bit. It probably doesn't matter now, but if we added more stuff later, we could be losing connectivity before the last ones complete.